### ISSUE TITLE
Support for "lifetimes" in nested instances

### ISSUE THREAD

**Issue by SteVwonder** [flux-framework/flux-core#2371]
In the case of a nested instance, we should make sure that the duration of it's allocation is reflected in the "lifetime" of the resources in the nested instance's scheduler.  For example, the following should result in a submission failure in the sub-instance:

```
flux submit -t 1:00:00 flux start flux submit -t 2:00:00 app.exe
```

This could be extended to support nested instances that are launched by other resource managers using something like `libyogrt`.  For example, the follow should also be an error:

```
srun -t 1:00:00 flux start flux submit -t 2:00:00 app.exe
```

The former case may already be supported, and if so, the latter could solve some of the problems our users are having with `libyogrt` in nested flux instances.

One other consideration is: it would be nice for users to request the timelimit to be equal to the remaining lifetime of the enclosing instance.  Maybe we can use `-1` or `infinity` to represent this (e.g., `flux submit -t 1:00:00 flux start flux submit -t infinity app.exe`)

**Comment by SteVwonder** [comment]
> I'm guessing this should be done at the job-exec level

That is my gut instinct as well. Makes sense to me in terms of the "hand-off" of jobs between modules.  Once the job is running, it's in the hands of the exec system.

> Maybe this should be the default (i.e. if a user doesn't specify a timelimit then a system default value is chosen). The default expiration for generated R can be set to the minimum of now + default_timeout and current instance expiration (if any).

👍 

> As you suggest, validation of incoming job requests can emit an error if now + default_timeout > expiration. This would be part of feasibility checks (I don't remember our plan on how to support that)

Yeah, I think this should happen at ingest time as well as part of the scheduling loop. The scheduler should not (by default) schedule jobs that are guaranteed to not have the time they require.  That also would account for the time gap between ingest validation and the `alloc` request to the scheduler.

**Comment by teaguejt** [comment]
I've implemented something similar for Hermit since the service will attempt to assign jobs without regard for how much time a child actually has remaining. The implementation follows this process, which could probably be adapted for what you want:

1. When a child is created, it is passed its lifetime as a parameter and uses that to create an internal start time for future reference.
2. Every time the child receives a job to execute, it ensures it has enough time to complete the job based off the job's wall time and the child's remaining lifetime (plus a small safety factor, so an instance with one hour and one second remaining will not attempt to execute a one-hour job)
3a. If the child has enough lifetime remaining, it will execute the job
3b. If the child does NOT have enough lifetime remaining, it will return the job to the management service and then kill itself; the job goes to the front of the work queue and will be redistributed the next time another child asks for work

This solution seems to be working fairly well during my Hermit testing, although I just implemented it yesterday so it hasn't been fully examined or properly stressed yet.

One thing I'd like to take into consideration that this solution currently does not is the presence of a heterogeneous work queue, i.e. if a child rejects a job due to time constraints, try to find something else for it to do instead of just letting it kill itself.

**Comment by teaguejt** [comment]
There are a couple other cases that something like this would need to account for as well, I think.

- Jobs like MD jobs, where there is some minimum time to generate useful data and the upper time limit may be less important. I have yet to determine if the existing functionality can be used with different parameters or if this is something that requires its own implementation.
- For a large number of rejections, there can be a decent amount of overhead and traffic involved with my solution. I'm not sure how to mitigate that yet, but I'm thinking about it. It's certainly something that should be considered.

**Comment by grondo** [comment]
This was brought up in a slack discussion and it got me thinking of the following points:

 * We could trivially add a job validator plugin that checks that any _explicit_ duration set in jobspec would not exceed the current instance lifetime, and reject that job. However, the validator would not be able to do anything if no explicit duration is specified.
 * It would be nice to set a default time limit for jobs to the remaining instance lifetime, however, unless a job starts immediately, the default time limit chosen would be too long. 

Currently, [RFC 14](https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_14.html) states that:

>  A duration of 0 SHALL be interpreted as “unlimited” or “not set” by the system

Perhaps we should reword this to indicate that a `duration` of `0` indicates that the duration of the job is limited by the lifetime/duration of the enclosing instance. A scheduler would then set the `expiration` key to its own `expiration` (perhaps minus a shutdown period). This would allow the duration validator plugin to ignore jobs submitted with the default `duration` of `0`. It also solves the 2nd problem above, since the duration in the jobspec would not have to be edited by the system.

If the scheduler has to know the current instance lifetime anyway, it would also then be possible to reject jobs based on requested duration using the scheduler feasibility check instead of a separate plugin.

**Comment by garlick** [comment]
This makes complete sense to me.

**Comment by ofaaland** [comment]
SCR allows applications to write output datasets to cache, which are then flushed to the parallel file system.  This may be done periodically e.g. when an application wants to record state at multiple time steps to allow one to visualize changes over time.  

When output dataset K is written to cache, SCR attempts to determine whether enough time is left in the allocation to calculate + flush one more dataset (K+1).  If there is enough time, the application continues.  Otherwise, SCR halts the application, flushes output K to the parallel file system, and the job exits.  This prevents the application from "spending" time calculating data that will not be saved.

I believe this can only work if (a) the job itself tracks time limits and always sets the duration when creating a child job, or (b) libyogrt recursively queries all the ancestor jobs and returns the minimum remaining time it finds, or (c) flux propagates time limits to child jobs.  I could implement (b) but it seems like a temporary solution to me.

> The default expiration for generated R can be set to the minimum of now + default_timeout and current instance expiration (if any).

That seems like it would take care of SCR's needs.  It's also what I expected before experimenting, so it's what I think would not surprise users.

**Comment by dongahn** [comment]
FYI -- I think supporting this within fluxion will be a bit more involved the the ached-simple. So when you change the definition of the `duration` key, please open up a ticket against flux-sched.

**Comment by grondo** [comment]
PR against RFC 14 opened: flux-framework/rfc#315

**Comment by grondo** [comment]
This issue has mostly been addressed for flux-core only with the merge of PRs #4223 and #4224.

Still todo:

 * [ ] Support propagation of instance expiration in Fluxion (flux-framework/flux-sched#913)
 * [ ] Support propagation of time limit from foreign RM (perhaps using libyogrt, though this would create a circular dependency)

**Comment by grondo** [comment]
Maybe it would make sense to close this issue, since most of the work is done in flux-core, and open an issue to track the second bullet above:

> Support propagation of time limit from foreign RM (perhaps using libyogrt, though this would create a circular dependency)

**Comment by grondo** [comment]
Oops, that is already open as #4006. Closing this issue.
